llvm/lib/Analysis/TargetLibraryInfo.cpp

```diff
@@ -1459,19 +1459,16 @@ unsigned TargetLibraryInfoImpl::getWCharSize(const Module &M) const {
 }
 
 unsigned TargetLibraryInfoImpl::getSizeTSize(const Module &M) const {
-  // There is really no guarantee that sizeof(size_t) is equal to sizeof(int*).
-  // If that isn't true then it should be possible to derive the SizeTTy from
-  // the target triple here instead and do an early return.
-
-  // Historically LLVM assume that size_t has same size as intptr_t (hence
-  // deriving the size from sizeof(int*) in address space zero). This should
-  // work for most targets. For future consideration: DataLayout also implement
-  // getIndexSizeInBits which might map better to size_t compared to
-  // getPointerSizeInBits. Hard coding address space zero here might be
-  // unfortunate as well. Maybe getDefaultGlobalsAddressSpace() or
-  // getAllocaAddrSpace() is better.
-  unsigned AddressSpace = 0;
-  return M.getDataLayout().getPointerSizeInBits(AddressSpace);
+  // There is really no guarantee that sizeof(size_t) is equal to the index
+  // size of the default address space. If that isn't true then it should be
+  // possible to derive the SizeTTy from the target triple here instead and do
+  // an early return.
+
+  // Hard coding address space zero may seem unfortunate, but a number of
+  // configurations of common targets (i386, x86-64 x32, aarch64 x32, possibly
+  // others) have larger-than-size_t index sizes on non-default address spaces,
+  // making this the best default.
+  return M.getDataLayout().getIndexSizeInBits(/*AddressSpace=*/0);
```

Comment:
As suggested by @jrtc27 in https://github.com/llvm/llvm-project/pull/118729#discussion_r1876526965 I wonder if the cleanest option here (but best as a follow-up change) would be to introduce this address space for "generic C ABI pointer" (maybe `-C` or `-D`), which would allow avoiding the hardcoded zero here.

Reply:
Can we leave the local variable in to allow for that in future? It's not really doing any harm and just needlessly increases the diff.

Reply:
@jrtc27 The removal of the local was specifically requested by @arsenm in review. I don't feel strongly either way, if the two of you can agree on a preference.

Reply:
Removing it makes sense if it's going to stay as 0, but I don't think it does if it's going to become some longer expression (even if that may be a simple (member) function call) which IMO is cleaner to put in its own statement rather than having one long nested statement.

Reply:
Ping @arsenm

Reply:
I really don't see the plus of using the variable here. If a future change happens, it is no additional effort

Reply:
I don't have a particularly strong opinion here, I'm just generally of the view that it's better not to delete code you're going to reintroduce, but one line isn't a big deal.

Reply:
I'm making the executive decision to keep it the way it is in the PR currently.